File: scala/src/main/scala/Terraria.scala
package tadp.lenses.terraria

import monocle.macros.syntax.lens._
import monocle.macros.Lenses
import tadp.lenses.{Prism, Optional}

trait Armazon
case object Hierro extends Armazon
case object Oro extends Armazon
case object Plata extends Armazon

trait Material
case object Cristal extends Material
case object Vidrio extends Material
case object Polimero extends Material

trait Decoracion
case object Rubi extends Decoracion
case object Zafiro extends Decoracion
case object Esmeralda extends Decoracion

case class Ropa(superior: String, inferior: String, calzado: String, sombrero: String)

case class Lentes(material: Material, graduacion: Double) {
  def setGraduacion(graduacion: Double) = this.copy(graduacion= graduacion)
}

case class Gafas(armazon: Armazon, lentes: Lentes, decoracion: Decoracion) {
  def setLentes(lente: Lentes): Gafas = this.copy(lentes= lente)
  def setGraduacionGafas(graduacion: Double): Gafas = this.copy(
    lentes= lentes.copy(
      graduacion= graduacion
    )
  )
}

case class Personaje(nombre: String, gafas: Gafas, ropa: Ropa)

trait Item
case object Espada extends Item
case object Armadura extends Item

trait Cofre
case object CofreVacio extends Cofre
case class ContenidoCofre(item: Item) extends Cofre


object Test {
  val lentes = Lentes(Cristal, 2.0)
  val gafas = Gafas(Plata, lentes, Rubi)
  val ropaNormal = Ropa("Camisa", "Pantalon", "zapatillas", "")
  val personajePrisma = Prism[Personaje, String] {
    case Personaje(n, _, _) => Some(n)
    case _ => None
  }(nombre => Personaje(nombre, gafas, ropaNormal))


  val cofreConArmaduraOptional = Optional[Cofre, Item] {
    case cofre: ContenidoCofre => Some(cofre.item)
    case _ => None
  } { contenido => {
    case c: ContenidoCofre => c.copy(item = contenido)
    case cofre => cofre
  }
  }
}

File: haskell/src/Terraria.hs
{-# LANGUAGE Rank2Types, ExistentialQuantification #-}

module Terraria
  ()
where

{- 
  Estamos en el mundo de terraria! Tenemos personajes, que tienen un nombre y obviamente llevan ropa. La ropa está conformada por 
  la prenda superior (que puede ser una remera, una camisa o una armadura), la prenda inferior (un jean, un short o unas rodilleras),
  el calzado (unos zapatos, unas zapatillas o unas botas) y el sombrero (un fedora, un agorra o una corona).
  Además, los personajes son re cancheros y usan gafas, ya sea unos lentes de sol para tirar facha o unos anteojos con aumento para ver mejor.

  Adhiriendo a la filosofía de Heráclito: un personaje usa siempre la misma espada, pero va actualizando las medidas según lo necesite.
  Queremos hacer que un personaje pueda decorar su espada con un nuevo pomo.
-}

data Personaje = Personaje {
  nombre :: String,
  gafas :: Gafas,
  ropa :: Ropa
}

data Ropa = Ropa {
  superior :: String,
  inferior :: String,
  calzado :: String,
  sombrero :: String
}

data Gafas = Gafas {
  armazon :: Armazon,
  lentes :: Lentes,
  decoracion :: Decoracion
}

data Lentes = Lente {
  material :: Material,
  graduacion :: Double
}

data Armazon = Hierro | Oro | Plata
data Material = Cristal | Vidrio | Polimero
data Decoracion = Rubi | Zafiro | Esmeralda

setGraduacion :: Double -> Lentes -> Lentes
setGraduacion graduacion l = l { graduacion = graduacion }

setLentes :: Lentes -> Gafas -> Gafas
setLentes lentes g = g { lentes = lentes }

setGafas :: Gafas -> Personaje -> Personaje
setGafas gafas p = p { gafas = gafas }

setGraduacionGafas :: Double -> Gafas -> Gafas
setGraduacionGafas graduacion g = g { lentes = setGraduacion graduacion . lentes $ g }

cambiarPersonaje :: Double -> Personaje -> Personaje
cambiarPersonaje graduacion p = p { gafas = setGraduacionGafas graduacion . gafas $ p }

graduacionPersonaje :: Personaje -> Double
graduacionPersonaje = graduacion . lentes . gafas

lentesPersonaje :: Personaje -> Lente
lentesPersonaje = lentes . gafas

cambiarMaterialPersonaje :: Material -> Personaje -> Personaje
cambiarMaterialPersonaje nuevoMaterial p = 
  p { gafas = (gafas p) {
      lentes = (lentes $ gafas p) { 
        material = nuevoMaterial }
        }
    }





type Setter a b = (b -> b) -> a -> a

